src/Raven.Server/Documents/Patch/ScriptRunner.cs

```diff
@@ -1175,14 +1176,37 @@ public JsValue PutDocument(JsValue self, JsValue[] args)
                     }
                     reader = JsBlittableBridge.Translate(_jsonCtx, ScriptEngine, args[1].AsObject(), usageMode: BlittableJsonDocumentBuilder.UsageMode.ToDisk, removeSpecialMetadata: true);
 
-                    var putResult = _database.DocumentsStorage.Put(
-                        _docsCtx,
-                        id,
-                        _docsCtx.GetLazyString(changeVector),
-                        reader,
-                        //RavenDB-11391 Those flags were added to cause attachment/counter metadata table check & remove metadata properties if not necessary
-                        nonPersistentFlags: NonPersistentDocumentFlags.ResolveAttachmentsConflict | NonPersistentDocumentFlags.ResolveCountersConflict | NonPersistentDocumentFlags.ResolveTimeSeriesConflict
-                    );
+                    DocumentsStorage.PutOperationResults putResult;
+
+                    while (true)
+                    {
+                        try
+                        {
+                            putResult = _database.DocumentsStorage.Put(
+                                _docsCtx,
+                                id,
+                                _docsCtx.GetLazyString(changeVector),
+                                reader,
+                                //RavenDB-11391 Those flags were added to cause attachment/counter metadata table check & remove metadata properties if not necessary
+                                nonPersistentFlags: NonPersistentDocumentFlags.ResolveAttachmentsConflict | NonPersistentDocumentFlags.ResolveCountersConflict | NonPersistentDocumentFlags.ResolveTimeSeriesConflict
+                            );
+
+                            break;
+                        }
+                        catch (Voron.Exceptions.VoronConcurrencyErrorException)
+                        {
+                            // RavenDB-21091 / RavenDB-10581 - If we have a concurrency error on "doc-id/"
```

Comment:
Can we put this in some shared code? I assume that this is duplicated a bunch of times...